### PR TITLE
Fix mania SR inflation for hold note releases in quick succession

### PR DESCRIPTION
**Pull request by molneya**
This change fixes 2 critical issues with mania release holds:

- Hold releases can change difficulty depending on what column they are in (e.g. mirroring a map changes star rating in some circumstances)
- Hold releases in quick succession can inflate the difficulty as they are treated independently, despite how the player might actually play them

[Visual explanation for the new LN release nerf](https://www.desmos.com/calculator/7eboanfzne) created by @Eve-ning

Both of these changes should just be nerfing maps with holds and nothing else.

### CODE REVIEW

**Comment by stanriders** [comment]
Looks good from code and conceptual side for me. Can't really speak about how well that translates to gameplay tho, hope someone familiar with mania can take a look

**Comment by Kominaru** [comment]
> Can't really speak about how well that translates to gameplay tho, hope someone familiar with mania can take a look

Hi mania NAT here. Can confirm this fixes a lot of the issues with SR bloat caused from uneven releases, so all good on that end. 
We are currently looking onto whether this overnerfs some actually difficult LN patterns, buy in hindsight this is all good and benefitial.

**Comment by stanriders** [comment]
>We are currently looking onto whether this overnerfs some actually difficult LN patterns

As we briefly discussed with @molneya in pp dev discord, it's pretty much out of scope of this change anyway. Unless there is a quick way to balance them back, fixing LN maps can be done later in a different PR.

**Review by stanriders — DISMISSED** [review]

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs [hunk 1]
```diff
@@ -37,31 +37,33 @@ protected override double StrainValueOf(DifficultyHitObject current)
...
+                closestEndTime = Math.Min(closestEndTime, Math.Abs(endTime - holdEndTimes[i]));
+
                 // Decay individual strains
                 individualStrains[i] = applyDecay(individualStrains[i], current.DeltaTime, individual_decay_base);
             }
 
             holdEndTimes[column] = endTime;
 
+            // The hold addition only is valid if there is _no_ other note with the same ending. Releasing multiple notes at the same time is just as easy as releasing 1
+            holdAddition *= 1 / (1 + Math.Exp(0.5 * (24 - closestEndTime)));
```
Comment (abraker95) [inline osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs:65]: ## Code efficiency

As an optimization, this should be done only if `holdAddition` is more than 0.

The `24` looks like a magic number. It actually determines the point at which LN difficulty is nerfed by half. In this case that point is where note endings are 24 ms apart. There should really be a comment regarding this.

## Algorithm concerns

If I understood the code correctly, let's say `endTime` is the red shaded note's end time, and the end times of blue shaded notes are `holdEndTimes[i]`. Numbered below are columns `i`. 



Since the red shaded note and the note in column 2 are considered overlapping as per condition on line 49, `holdAddition` would then be `1.0`. Outlined in green are the two notes that have the smallest difference in their end times, which would make `closestEndTime` be 0.  Therefore, `holdAddition` would also be 0.

In short, any synced hold note ends are going to undermine any other hold note ends that have a significant time difference. This extends to the matter that this does not take into account inter-hand patterns. Two note with synced ends on different hands can nerf notes that require significant hand dexterity on single hand.

This doesn't look desirable. The red shaded note is requires non 0 finger dexterity to perform, but this is nerfing it to 0.

As a note for the future in regards to balancing difficulty of LN patterns, I also believe this should be OD dependent.

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs [hunk 2]
```diff
@@ -37,31 +37,33 @@ protected override double StrainValueOf(DifficultyHitObject current)
...
             for (int i = 0; i < holdEndTimes.Length; ++i)
             {
-                // If there is at least one other overlapping end or note, then we get an addition, buuuuuut...
+                // If there is at least one other overlapping end or note, then we get an addition
                 if (Precision.DefinitelyBigger(holdEndTimes[i], maniaCurrent.StartTime, 1) && Precision.DefinitelyBigger(endTime, holdEndTimes[i], 1))
                     holdAddition = 1.0;
 
-                // ... this addition only is valid if there is _no_ other note with the same ending. Releasing multiple notes at the same time is just as easy as releasing 1
-                if (Precision.AlmostEquals(endTime, holdEndTimes[i], 1))
-                    holdAddition = 0;
```
Comment (abraker95) [inline osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs:53]: Just going to comment why this change was needed. 

The for loop iterated through notes across the columns one-by-one which caused the order of the notes matter when it came to these two checks. This gave the notes on the right side of the stage higher priority than the ones on the left, effectively a "race condition" between the two checks in setting `holdAddition`.

An alternative solution would have been to split this into two `for` loops but ¯\\\_(ツ)\_\/¯

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs (hunk 1 above)
Comment (molneya) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs]: > As an optimization, this should be done only if holdAddition is more than 0.

Done

> The 24 looks like a magic number.

Commented its purpose here.

>In short, any synced hold note ends are going to undermine any other hold note ends that have a significant time difference. 

Totally agree with you on this, but I didn't want to change the original algorithm completely. The main purpose here was to make sure that mirroring the map didn't change the star rating. Taking what hand the note is on should definitely be taken into account in future fixes.

> I also believe this should be OD dependent.

Definitely, but for ranked maps this shouldn't be too much of an issue as the values for OD are not too different.

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs (hunk 2 above)
Comment (bdach) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs:53]: Calling this bug a "race condition" is a stretch even for air quotes. There's no threading here.

Comment (abraker95) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs:53]: I don't know how else to describe it, so I chose the the closest analogy. From a algorithmic pov, the intention was that the following checks should be executed sequentially. Instead, they are conflicting, kinda executing at same time, albeit in an interweaved order.
```
if any(note are overlapping)
   holdAddition = 1.0;

if any(note have same end times)
   holdAddition = 0;
```

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs (hunk 1 above)
Comment (Richardlx) [inline osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs]: Apologies if I misunderstood the context here. So with this we basically nullify any `holdAddition` to a hold whenever there is a "nearby" (in terms of timing, its 24ms which is about 1/12 under 190 bpm; I am not sure if anyone sets this time with any map specifically in mind but otherwise it does feel like a magic number here... but whatever) hold also released. That's fine. 

My bigger question is, imagine in a hypothetical 7k scenario let's say in a LN wall pattern we have holds at column 1, 2, 3, 4, 5, 6, 7 and we have a release at 5. Cool, the note gets its bonus as it should: releasing a hold within a LN wall is not easy. Now consider a different pattern: we have a release at column 5 and 6. Now due to this line both notes lose their bonus, while practically speaking the pattern is not any easier. 

I am not saying current calculation does not have the same problem. However since you are now nerfing a lot of these scenarios, I do think this changes will actually make some actually hard but underrated LN patterns even more underrated (let's say a pattern consisting two long holds on column 4/7 with streams of double/triple shorter holds on other columns. A lot of these shorter holds might lose their bonus).

I feel like maybe we should consider scaling `holdAddition` bonus to how many holds are currently being pressed, not making it a boolean solely rely on an arbitrary `closestEndtime` (since this literally means we only consider bonus by looking at the next holds, while potentially we can have 4~8 LNs going on at the same time... ).  I know it is probably beyond the scope of this PR, but given how much this change may affect rating for all LN patterns I think at least we should reconsider our approach

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs [hunk 3]
```diff
@@ -37,31 +37,35 @@ protected override double StrainValueOf(DifficultyHitObject current)
             var maniaCurrent = (ManiaDifficultyHitObject)current;
             double endTime = maniaCurrent.EndTime;
             int column = maniaCurrent.BaseObject.Column;
+            double closestEndTime = 100;
```
Comment (Eve-ning) [inline osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs:41]:
```suggestion
bool isOverlapping = False:
```

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs (hunk 1 above)
Comment (abraker95) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs]: See my review [here](https://github.com/ppy/osu/pull/17913#discussion_r855365900). It basically says the same thing, with @molneya addressing the concern.

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs [hunk 4]
```diff
@@ -37,31 +37,35 @@ protected override double StrainValueOf(DifficultyHitObject current)
...
             double holdFactor = 1.0; // Factor to all additional strains in case something else is held
             double holdAddition = 0; // Addition to the current note in case it's a hold and has to be released awkwardly
 
             // Fill up the holdEndTimes array
             for (int i = 0; i < holdEndTimes.Length; ++i)
             {
-                // If there is at least one other overlapping end or note, then we get an addition, buuuuuut...
+                // If there is at least one other overlapping end or note, then we get an addition
                 if (Precision.DefinitelyBigger(holdEndTimes[i], maniaCurrent.StartTime, 1) && Precision.DefinitelyBigger(endTime, holdEndTimes[i], 1))
                     holdAddition = 1.0;
```
Comment (Eve-ning) [inline osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs:50]: Simplify this since the `1.0` is not important when L66 compares `>0` anyways.
Making this any value `>0` would trigger the same result

`isOverlapping` would be more explicit

```suggestion
                isOverlapping = (Precision.DefinitelyBigger(holdEndTimes[i], maniaCurrent.StartTime, 1) && Precision.DefinitelyBigger(endTime, holdEndTimes[i], 1));
```

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs [hunk 5]
```diff
@@ -37,31 +37,35 @@ protected override double StrainValueOf(DifficultyHitObject current)
...
+
                 // Decay individual strains
                 individualStrains[i] = applyDecay(individualStrains[i], current.DeltaTime, individual_decay_base);
             }
 
             holdEndTimes[column] = endTime;
 
+            // The hold addition only is valid if there is _no_ other note with the same ending. Releasing multiple notes at the same time is just as easy as releasing 1
+            // Nerfs the hold addition by half if the closest release is 24ms away
+            if (holdAddition > 0)
```
Comment (Eve-ning) [inline osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs:66]: As per the above suggestions, this will be more explicit
```suggestion
            if (isOverlapping)
```

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs [hunk 6]
```diff
@@ -37,31 +37,35 @@ protected override double StrainValueOf(DifficultyHitObject current)
             var maniaCurrent = (ManiaDifficultyHitObject)current;
             double endTime = maniaCurrent.EndTime;
             int column = maniaCurrent.BaseObject.Column;
+            double closestEndTime = 100;
```
Comment (Eve-ning) [inline osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs:40]: Is there anyway to make this `Double.MaxValue` ? or avoid a `100`? It's not explicit why it's 100

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs (hunk 1 above)
Comment (molneya) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs]: > I am not sure if anyone sets this time with any map specifically in mind but otherwise it does feel like a magic number here

The 24ms value was chosen as its about 3/4 of the 300G timing window, still leaving about 1/4 of the window to hit the notes perfectly as if you were playing them on the same timing. The exact number is slightly arbitrary, but this value gave a good balance between nerfing overrated patterns while still keeping the bonuses when they really are awkward.

> ... we have a release at column 5 and 6. Now due to this line both notes lose their bonus, while practically speaking the pattern is not any easier.

The original algorithm had this same problem. The purpose of this change is to make sure that no matter the ordering of the columns the end star rating will be the same. If multiple notes share the same release timing, you can think of it as them all sharing the same `holdAddition`, as it's applied because of column 1. Or in other words, a new `holdAddition` will be given for every new snap a hold end appears (when it meets the awkwardness conditions). Addressing issues with the `holdAddition` itself should probably be addressed in a different PR.

**Comment by smoogipoo** [comment]
SR/PP changes sheet: https://docs.google.com/spreadsheets/d/1vvRUBze0cK8qBVcT0ywHyhLt3woZLz67j4HPKLmuJ_A/edit?usp=sharing

**Comment by smoogipoo** [comment]
Since we don't really have a pp committee for mania, I'm going to rely on approvals from @abraker95 @Eve-ning and @Kominaru for merge.

**Comment by Kominaru** [comment]
> Since we don't really have a pp committee for mania, I'm going to rely on approvals from @abraker95 @Eve-ning and @Kominaru for merge.

All good on my end! 
On that note, I think it would be a good idea to actually conform a formal pp committee knowing all four of us are pretty much among the pack of people that are actively contributing to pp dev

**Review by WindowsMeosu — DISMISSED** [review]

**Review by abraker95 — DISMISSED** [review]
looks good

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs [hunk 7]
```diff
@@ -37,31 +37,35 @@ protected override double StrainValueOf(DifficultyHitObject current)
...
                 // Decay individual strains
                 individualStrains[i] = applyDecay(individualStrains[i], current.DeltaTime, individual_decay_base);
             }
 
             holdEndTimes[column] = endTime;
 
+            // The hold addition is given if there was an overlap, however it is only valid if there are no other note with a similar ending.
+            // Releasing multiple notes is just as easy as releasing 1. Nerfs the hold addition by half if the closest release is 24ms away.
+            if (isOverlapping)
+                holdAddition = 1 / (1 + Math.Exp(0.5 * (24 - closestEndTime)));
```
Comment (Eve-ning) [inline osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs:67]: A final code touch up will be to move the `24` as a `private const` variable.
I don't have recommendations of a good variable name however

you may want to also include my ASCII graph art here
```
// hold_addition
//     ^
// 1.0 + - - - - - -+-----------
//     |           /
// 0.5 + - - - - -/   Sigmoid Curve
//     |         /|
// 0.0 +--------+-+---------------> Release Difference / ms
//          <CONST_NAME>
```

Comment (molneya) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs]: Nice ASCII art. Included.

**Review by Eve-ning — APPROVED** [review]
Looks good, while it may underrate some LN maps, this is an essential step forward as it corrects unexpected behavior from `hold_addition`, good job

**Review by smoogipoo — APPROVED** [review]
Lgtm, thanks!

Not yet sure when the next diffcalc release will be due to changes in the infrastructure.
